util/logtools/encounter_tools.ts

```diff
@@ -0,0 +1,313 @@
+      if (this.skipZone()) {
+        this.initializeZone();
+        return;
+      }
+
+      this.currentZone.name = cZ.groups.name;
+      this.onStartZone(line, this.currentZone.name, cZ.groups);
+      return;
+    }
+
+    if (this.skipZone() || !this.currentZone.name)
+      return;
+
+    const cW = this.regex.cactbotWipe.exec(line);
+    if (cW?.groups) {
+      if (this.currentFight.startTime)
+        this.initializeFight();
+      return;
+    }
+
+    const wipe = this.regex.wipe.exec(line);
+    if (wipe?.groups) {
+      if (this.currentFight.startTime)
+        this.initializeFight();
+      return;
+    }
+
+    const win = this.regex.win.exec(line);
+    if (win?.groups) {
+      if (this.currentFight.startTime) {
+        this.haveWon = true;
+        this.initializeFight();
+      }
+      return;
+    }
+
+    if (!(this.currentFight.startTime || this.haveWon || this.haveSeenSeals)) {
+      let a = this.regex.playerAttackingMob.exec(line);
+      if (!a)
+        a = this.regex.mobAttackingPlayer.exec(line);
+      if (a?.groups) {
+        this.onStartFight(line, this.currentZone.name, a.groups);
+        return;
+      }
+    }
+
+    for (const regex of this.sealRegexes) {
+      const s = regex.exec(line);
+      if (s?.groups?.name) {
+        this.haveSeenSeals = true;
+        this.currentSeal = s.groups?.name;
+        this.onStartFight(line, this.currentSeal, s.groups);
+        return;
+      }
+    }
+
+    for (const regex of this.unsealRegexes) {
+      const u = regex.exec(line);
+      if (u) {
+        this.currentSeal = null;
+        if (this.currentFight)
+          this.initializeFight();
+        return;
+      }
+    }
+  }
+
+  // start/endZone always bracket all fights and seals.
+  // All starts and ends of the same type are ordered and do not nest.
+  // Fights and seal start/end may interleave with each other.
+  // TODO: probably this should follow an "event bus" model instead of requiring derived classes.
+  onStartZone(line: string, name: string, matches: NetMatches['ChangeZone']): void {
+    this.currentZone = {
+      name: name,
+      startLine: line,
+      zoneId: parseInt(matches.id),
+      startTime: this.dateFromMatches(matches),
+    };
+  }
+  onStartFight(line: string, name: string, matches: NetAnyMatches): void {
+    this.currentFight = {
+      name: name,
+      startLine: line,
+      startTime: this.dateFromMatches(matches),
+    };
+  }
+
+  dateFromMatches(matches: NetAnyMatches): Date {
+    // No current match definitions are missing a timestamp,
+    // but in the event any are added without in future, we will be ready!
+    if (!matches.timestamp)
+      throw new UnreachableCode();
+    return new Date(Date.parse(matches.timestamp));
+  }
+}
+
+export class EncounterCollector extends EncounterFinder {
+  zones: Array<ZoneEncInfo>;
+  fights: Array<FightEncInfo>;
+  lastZone: ZoneEncInfo;
+  lastFight: FightEncInfo;
+  lastSeal: string | null;
```

Comment:
Same thing with `undefined` vs `null` here, as well as initializing members in their declaration.